### PR TITLE
[FSDP] `summon_full_params()` in computation stream

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#86836 [FSDP] `summon_full_params()` in computation stream**
* #86833 [FSDP] Rename streams
* #87308 [FSDP][2/N] Fix grad zero vs. `None` edge case
* #87314 [FSDP][1/N] Update `summon_full_params(with_grads)` `None` gradient

This should help with memory usage. In particular, this allows FSDP to use caching allocator blocks from the computation stream for the `summon_full_params()` all-gathers, which should help avoid over-allocating blocks to the unshard stream.

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -3063,8 +3065,9 @@ def _summon_full_params(
 
         self._clear_grads_if_needed()
         free_unsharded_flat_params = [handle.needs_unshard() for handle in self._handles]
-        self._unshard(self._handles)
-        self._streams["computation"].wait_stream(self._streams["unshard"])
+        # No need to call `wait_stream()` since we unshard in the computation
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3068]: Curious, why would we want to move this to computation stream?

**Review by rohan-varma — APPROVED** [review]
LGTM, but curious why we need this change.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:3068]: This allows us to use caching allocator blocks from the computation stream for these all-gathers, which should help avoid over-allocating blocks to the unshard stream.

**Pull request closed without merging**